encoding: Return to base-100 encoding for Decimal

nvb opened the pull request:
cc. @bdarnell 

Closes #5965.

Again unfortunately this is a breaking change. I discussed with Ben whether
we should provide a migration path

The previous encoding approach for DECIMAL was broken in two ways:
- It prefixed the mantissa with the mantissa's length, which meant that
  decimal values did not always sort correctly. For instance `.01111`
  would sort before `0.02`.
- It encoded its exponent in a base that was not a factor of its
  mantissa's base, which also permitted very subtle ordering errors.
  Here, the encoding's exponent was base-10 encoded while
  its mantissa was base-256 encoded. This was not a problem in the
  current iteration of the previous encoding because the mantissa length
  varint (discussed above) would prevent any strange ordering. However, if
  we had moved from the mantissa length prefix approach to the escaped
  0x00 byte approach for encoding the mantissa, this issue would have
  come up. In practice, it means that values like `255 = [0xff]` would
  sort before `256 = [0x01 0x00]`. It is critical that the exponent and
  the mantissa share the same base (or the exponent is a factor of the
  mantissa's base) so that the exponent will increment whenever the
  mantissa's byte size grows.

I spent all day trying to convert the base-10 exponent into a base-256 (or
base-16) exponent so we could avoid marshaling through strings on encoding
and decoding, but had little luck. If anyone has ideas, I'd love to hear
them.

The current approach uses a base-100 exponent and a base-100 mantissa.
It is similar to the first encoding of decimal, with a few key
differences.
- The null terminator is searched for AFTER the uvarint has been
  decoded. This was an issue with both the previous float and decimal
  encodings, which both assumed that the exponent uvarint would be free
  of 0x00 bytes. In those instances, there was a risk of searching for
  the terminator, only to find a 0x00 byte in the uvarint.
- A few performance tweaks were made to reduce an allocation from
  encoding and to reduce an allocation from decoding.
- We're now using the `{Encode/Decode}Uvarint{Ascending/Descending}`
  functions instead of the old uvarint functions.
- I've added a random encoding test to make sure this is thoroughly correct. 
  So far, so good...

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5994)

<!-- Reviewable:end -->

nvb committed:
encoding: Return to base-100 encoding for Decimal

The previous encoding approach for DECIMAL was broken in two ways:
- It prefixed the mantissa with the mantissa's length, which meant that
  decimal values did not always sort correctly. For instance `.01111`
  would sort before `0.02`.
- It encoded its exponent in a base that was not a factor of its
  mantissa's base, which also permitted very subtle ordering errors.
  Here, the encoding's exponent was base-10 encoded while
  its mantissa was base-256 encoded. This was not a problem in the
  current iteration of the previous encoding because the mantissa length
  varint (discussed above) would prevent any strange ordering. However, if
  we had moved from the mantissa length prefix approach to the escaped
  0x00 byte approach for encoding the mantissa, this issue would have
  come up. In practice, it means that values like `255 = [0xff]` would
  sort before `256 = [0x01 0x00]`. It is critical that the exponent and
  the mantissa share the same base (or the exponent is a factor of the
  mantissa's base) so that the exponent will increment whenever the
  mantissa's byte size grows.

I spent all day trying to convert the base-10 exponent into a base-256
exponent so we could avoid marshalling through strings on encoding and
decoding, but had little luck. If anyone has ideas, I'd love to hear
them.

The current approach uses a base-100 exponent and a base-100 mantissa.
It is similar to the first encoding of decimal, with a few key
differences.
- The null terminator is searched for AFTER the uvarint has been
  decoded. This was an issue with both the previous float and decimal
  encodings, which both assumed that the exponent uvarint would be free
  of 0x00 bytes. In those instances, there was a risk of searching for
  the terminator, only to find a 0x00 byte in the uvarint.
- A few performance tweaks were made to reduce an allocation from
  encoding and to reduce an allocation from decoding.
- We're now using the `{Encode/Decode}Uvarint{Ascending/Descending}`
  functions instead of the old uvarint functions.
- I've added a random encoding test to make sure this is correct. So far
  so good.

knz commented:
Woah using base-100 is a huge step backwards in terms of efficiency. 
Each digit encodes in 6.64 bits which is 83% storage efficiency per byte.

What about base 160, which encodes in bytes at 91%, or base 40960 which encodes in digits of 16 bits at 96% efficiency?

petermattis commented on roachpb/data.go:
```diff
@@ -446,9 +446,9 @@ func (v *Value) SetDuration(t duration.Duration) error {
-	v.RawBytes = encoding.EncodeDecimalAscending(v.RawBytes, dec)
+	encodingSizeOverestimate := headerSize + encoding.OverestimateDecimalSize(dec)
+	v.RawBytes = make([]byte, encodingSizeOverestimate)
+	v.RawBytes = encoding.EncodeDecimalAscending(v.RawBytes[:headerSize], dec)
```
The values stored in `Value` don't need to be sorted. I wonder if there should be a different encoding for decimals here. In particular, we could keep the old encoding (which didn't sort properly) but was significantly faster to encode/decode.

nvanbenschoten replied:
Yeah good point. I added back a `EncodeNonSortingDecimal` function to use for `roachpb.Value`. Because this is not for key encoding, there were a few places I could gain some speed. They included cutting out the length/terminator from the encoding itself because the length of the `Value.dataBytes` is already known.

nvb committed:
util/encoding: Use slimmer old encoding for non-sorting decimal encoding

This change preserves the old encoding for encoding decimals where the
encoded byte slices do not need to sort lexicographical. Specifically,
we can use this to encode `roachpb.Values`.

nvanbenschoten commented:
The result of this change on performance is along the lines of:

```
name               old time/op    new time/op    delta
ValueSetDecimal-4     199ns ± 2%     181ns ± 1%    -9.17%        (p=0.000 n=20+10)
ValueGetDecimal-4     218ns ± 2%     196ns ± 1%   -10.07%        (p=0.000 n=20+10)
EncodeDecimal-4       155ns ± 3%     396ns ± 2%  +156.47%        (p=0.000 n=10+10)
DecodeDecimal-4       206ns ± 2%     504ns ± 1%  +143.96%         (p=0.000 n=10+9)

name               old alloc/op   new alloc/op   delta
ValueSetDecimal-4     32.0B ± 0%     32.0B ± 0%      ~     (all samples are equal)
ValueGetDecimal-4     96.0B ± 0%     96.0B ± 0%      ~     (all samples are equal)
EncodeDecimal-4      0.00B ±NaN%    69.00B ± 0%     +Inf%        (p=0.000 n=10+10)
DecodeDecimal-4       96.0B ± 0%    128.0B ± 0%   +33.33%        (p=0.000 n=10+10)

name               old allocs/op  new allocs/op  delta
ValueSetDecimal-4      1.00 ± 0%      1.00 ± 0%      ~     (all samples are equal)
ValueGetDecimal-4      2.00 ± 0%      2.00 ± 0%      ~     (all samples are equal)
EncodeDecimal-4       0.00 ±NaN%      2.00 ± 0%     +Inf%        (p=0.000 n=10+10)
DecodeDecimal-4        2.00 ± 0%      3.00 ± 0%   +50.00%        (p=0.000 n=10+10)
```

Note that decimal key encoding got slower but decimal value encoding actually got faster (and smaller).

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[util/encoding/decimal_test.go, line 220 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEtmC5nBXhQZiErSAc:-KFEtmC5nBXhQZiErSAd:-1410151456) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/decimal_test.go#L220)):*
How long does this test take? I would bump this up until the test takes ~1s.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFEu9ayzZwHMNIyvG2M:1724353026) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
Uh-oh! We don't want to change `bytesMarker` and all of the other markers. Doing so will break compatibility for all encoded data.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Let's add another test which encodes a value for each of the 

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFEuwwtogQvbZbSK__q:552054498) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
Err, never mind. I see that you're just hardcoding the value instead of deriving it from the `decimal` markers. Carry on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-:-KFEuXBOpUMQvZADzixG:-1121374696)*
<!-- Sent from Reviewable.io -->

petermattis commented:
My comment was cut off. I was going to suggest that we add a test that uses each of the encoding routines, hashes (sha1) the output and verifies against golden data, but we more or less have that with the various unit tests.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-:-KFEv3xTDE_E4otKwd2Z:47817054)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-:-KFEvOQr4s8cZhJCBm0P:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFFJCWL_ohCJrn9XDD1:1964836411) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
I think it's fine to move towards hard-coding constants here (we don't want to make these values easy to change accidentally), but there should be a comment about this gap in the sequence (and the gap between `durationBigPosMarker` and `decimalNan`. why not start `decimalNan` at `durationBigPosMarker+1`?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[util/encoding/decimal_test.go, line 220 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEtmC5nBXhQZiErSAc:-KFFvtXrZgFzkB7NG3Ap:1084272768) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/decimal_test.go#L220)):*
I bumped both of the random tests to take about a second.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFFw6HEr6x_hMFN4dyo:512491286) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
Added a comment.

And I was hoping to leave some room for "growth". The reason I had to move the markers around here was because there was no room to add an extra tag. I dont feel super strongly about this though. Do you think I should push the decimal markers up against `durationBigPosMarker`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFG3SQD3juvjjmmxCZh:-1105562055) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
I'm not sure. I don't feel strongly either way. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFG2ZJkIK6PnwSpldml:1773495247) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
I'm thinking about fragmentation: this change relied on the fact that there was a large gap between `durationBigPosMarker` and `IntMin`. You've left a small gap which gives duration or decimal a little room to expand, but at the expense of space in the larger unclaimed area which might be needed for future data types that need a chunk of contiguous space. I don't feel strongly about it but I'd prefer to pack the types closely. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->

nvb committed:
encoding: Remove fragmentation between encoding prefix markers

This change gets rid of the gap between the duration and decimal prefix
markers in an effort to reduce fragmentation.

nvanbenschoten commented:
Reviewed 1 of 3 files at r2.
Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[util/encoding/encoding.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994#-KFEu9ayzZwHMNIyvG2L:-KFGBiT81V8GoMJSUJLn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/880bf90c4958c2daa8fe293e80bf97f2190f8ab3/util/encoding/encoding.go#L48)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5994)*
<!-- Sent from Reviewable.io -->
